[GridBundle] Remove unsupported drivers during container compilation

dantleech opened the pull request:
| Q               | A
| --------------- | ---
| Bug fix?        | -
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | #4809 
| License         | MIT

This PR allows the drivers in the GridBundle to be explicitly enable.

The default behavior is to enable both dbal and orm drivers:

```yaml
sylius_grid:
    drivers: [ "doctrine/orm", "doctrine/dbal" ]
```

dantleech commented on src/Sylius/Bundle/GridBundle/DependencyInjection/Configuration.php:
```diff
@@ -38,6 +39,20 @@ public function getConfigTreeBuilder()
+    {
+        $node
+            ->children()
+                ->arrayNode('drivers')
```
Not sure how ambiguous this is -- I assume the user could equally register their own drivers, in whch case they would not need to include them here.

Should add an `->info` element.

dantleech commented:
Or maybe it would be better to use either `class_exists` or a compiler pass to activate the drivers.

dantleech commented:
@pjedrzejewski do you think this would be best done as a compiler pass? e.g. if the `doctrine.orm.entuty_manager` exists, then add the `doctrine/orm` driver?

pjedrzejewski commented:
There is a chance we get into trouble with compiler pass, but that would be nicer DX indeed. Let's try it!

dantleech commented:
Updated. I updated the `RegisterDriverPass` to check the dependencies/references of the driver definitions, if the reference is not available in the container, then the driver will be removed (and not added to the registry).

This is still slightly bad DX as it hides the avialability of the drivers - 

An alternative could be to replace the invalid definitions with an `UnsupportedDriver` class which would throw an exception upon instantiation asking the user to install the required dependencies (e.g. doctrine ORM, PHPCR ODM). But I feel that might be over-engineering for this specific bundle.

pjedrzejewski commented:
I agree that removing the services without notice to the developer can be a bit confusing. If someone adds a custom driver and make a typo or reference wrong service, his driver will not appear. So maybe more explicit would be your initial solution where you define enabled drivers in yaml. At least for the ones supplied by SyliusGridBundle default, all custom drivers should be loaded automatically. I'd just split ``drivers.xml`` and put doctrine/orm and doctrine/dbal in separate files and load them or not based on this configuration, wdyt?

pamil commented:
In the end, I think the best solution for both ResourceBundle and GridBundle drivers is to make the drivers like [security authentication listeners in Symfony SecurityBundle](http://symfony.com/doc/current/cookbook/security/custom_authentication_provider.html#the-factory) or [configuration sources in Sylius ThemeBundle](http://docs.sylius.org/en/latest/bundles/SyliusThemeBundle/configuration_sources/creating_custom.html). The one I've written for ThemeBundle only modifies the container if it was activated, so you can enforce having some dependencies while there's still possibility to disable problematic driver.

pamil commented on src/Sylius/Bundle/GridBundle/Tests/DependencyInjection/Compiler/RegisterDriversPassTest.php:
```diff
@@ -0,0 +1,97 @@
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+class RegisterDriversPassTest extends \PHPUnit_Framework_TestCase
```
This lib makes testing compiler passes & extensions a lot easier - https://github.com/matthiasnoback/SymfonyDependencyInjectionTest#testing-a-compiler-pass :)

dantleech replied:
tbh, I used to use that library myself but over time I (personally) found that it was just obfuscating the tests. imo it's not worth abstracting.

pjedrzejewski commented:
@pamil Yeah, that would be the best, but will require a lot of work now, with not that big benefit currently. Let's start with something simpler to implement. :)

dantleech commented:
I also like the "provider" approach - it means that all the "drivers" will be available (i.e. if the user makes a typo it means that we can show them all the correct "answers" - good discoverability) and the providers can also throw good exceptions (i.e. "You need to composer require "doctrine/phpcr-odm") in order to use the `doctrine/phpcr-odm` driver.

On the other hand, by explicitly enabling the "drivers" we oblige the user to do this for every single bundle which has drivers - which is not such good DX.

But also don't mind implementing the explicit solution for now.

dantleech commented:
Replaced by https://github.com/Sylius/Sylius/pull/4972
